File: LinkedListAlgo_Python/linkedlist.py
class Node:
    """Node class for representing individual nodes in the linked list.
    """
    def __init__(self, data, next):
        self.data = data
        self.next = next


class LinkedList:
    """"LinkedList class for storing a list of nodes.
    """
    head = None
    count = 0

    def add_at_end(self, data):
        """Adds a node at the end of the list.
        """
        if self.head is None:
            self.head = Node(data, None)
            self.count += 1
            return
        
        current_node = self.head
        while current_node.next:
            current_node = current_node.next

        current_node.next = Node(data, None)
        self.count += 1
        return
    
    def add_at_start(self, data):
        """Adds a new node at the start of the list.
        """
        if self.head is None:
            self.head = Node(data, None)
            self.count += 1
            return
        
        node = Node(data, self.head)
        self.head = node 
        self.count += 1
        return

    def remove_from(self, index):
        """Removes a node at the specified index from the list.
        """
        if index < 0 or index > self.list_count():
            raise ValueError("Index out of range")
        
        current_node = self.head
        previous_node = None
        count = 0
        while current_node:
            if count == index - 1:
                # Check if node to be removed is the head
                if index -1 == 0:
                    if current_node.next is None:
                        self.head = None
                        self.count -= 1
                    else:
                        self.head = current_node.next
                        self.count -= 1
                    return

                # If tail is to be removed
                if current_node.next is None:
                    previous_node.next = None
                    self.count -= 1
                else:
                    # Node other than head or tail is to be removed
                    previous_node.next = current_node.next
                    self.count -= 1
                return
            
            previous_node = current_node
            current_node = current_node.next
            count += 1

    def list_count(self):
        """Keeps track of the number of nodes in the list.
        """
        return self.count

    def print_info(self):
        """Prints the nodes in the list.
        """
        current_node = self.head
        linkedlist_data = ""

        while current_node:
            linkedlist_data += f"{current_node.data}-->"
            current_node = current_node.next
        
        return linkedlist_data

    def reverse(self):
        """Reverse the LinkedList.
        """

        previous_node = None

        while self.head != None:
            next_node = self.head.next
            self.head.next = previous_node
            previous_node = self.head
            self.head = next_node
        
        self.head = previous_node

    def find_middle_node(self):
        fast_ptr = self.head
        slow_ptr = self.head
        previous_slow_ptr = slow_ptr

        while (fast_ptr != None) and (fast_ptr.next != None):
            fast_ptr = fast_ptr.next.next
            previous_slow_ptr = slow_ptr
            slow_ptr = slow_ptr.next
        
        # If fast_ptr is None, then list is even
        # If fast_ptr.next is None, then list is odd
        if fast_ptr is None:
            print(f"Middle nodes are {previous_slow_ptr.data} and {slow_ptr.data}")
        else:
            print(f"Middle node is {slow_ptr.data}")
    
    def has_loop(self):
        # Using Floyd's Cycle-Finding Algorithm
        fast_ptr = self.head
        slow_ptr = self.head

        while (fast_ptr != None) and (fast_ptr.next != None):
            slow_ptr = slow_ptr.next
            fast_ptr = fast_ptr.next.next

            if slow_ptr == fast_ptr:
                return True
        
        return False
    
    def create_loop(self):
        current_node = self.head

        while current_node.next:
            current_node = current_node.next
        
        # Point the last node's next to the head.
        current_node.next = self.head

File: LinkedListAlgo_Java/LinkedList.java
public class LinkedList{
  private class Node{
    private int value;
    private Node next;

    private Node(int value){
      this.value = value;
    }
  }

  private Node first;
  private Node last;

  private boolean isEmpty(){
    return first == null;
  }

  public void addLast(int item){  
    var node = new Node(item);
    if(isEmpty()){
      first = last = node;
    }else{
      last.next = node;
      last = node;
    }
  }

  public void addFirst(int item){
    var node = new Node(item);
    if(isEmpty()){
      first = last = node;
    }else{
      node.next = first;
      first = node;
    }    
  }

  public void deleteFirst(){
    if(isEmpty()){
      System.out.println("LinkedList is empty");
    }else if(first == last){
      first = last = null;
      return;
    }
    else{
      var second = first.next;
      first.next = null;
      first = second;
    }
  }

  public int indexOf(int item){
    int count = 0;
    Node current = first;
    while(current != null){
      if (current.value == item) return count;
      current = current.next;
      count += 1;
    }
    return -1;
  }

  // Implement the more efficient way, current implementation is o(n)
  // Find a way to make it o(1) by storing count of items whenever
  // items are added and removed from the list.
  public int len(){
    int count = 0;
    Node current = first;
    while(current != null){
      current = current.next;
      count += 1;
    }
    return count;
  }

  public boolean contains(int item){
    Node current = first;
    while(current != null){
      if (current.value == item) return true;
      current = current.next;
    }
    return false;
  }

  public void deleteLast(){
    if(isEmpty()){
      System.out.println("LinkedList is empty");
    }else if(first == last){
      first = last = null;
    }
    else{
      Node current = first.next;
      Node previous = first;
      while(current != null){
        if(current.next == null){
          last = previous;
          last.next = null;
          return;
        }
        previous = current;
        current = current.next;
      }
    }
  }

  public void reverse(){
    // Store first current value as first
    Node current = first;
    Node previous_node = null;
    while(current != null){
      Node next_node = current.next;

      // if linkedList has only one node
      // do nothing
      if(current == last){
        return;
      }
      // check if next node is null (if last node is found)
      // set that node as first and it's next node as prev node
      else if(next_node == null){
        current.next = previous_node;
        first = current;
      }
        // if first node is found
        // set that node as last node and 
        // set it's next node as null
      else if(current == first){
        current.next = null;
        last = current;
      }
        // if node is in the middle
        // set it's next node to the prev node
      else{
        current.next = previous_node;
      }
      previous_node = current;
      current = next_node;     
    }
  }

  public void printMiddle(){
    // check if it's empty
    if(isEmpty()){
      System.out.println("List is empty");
      return;
    }

    // set both slow and fast pointer to first node
    Node slow_ptr = first;
    Node fast_ptr = first;

    // Check if both fast_ptr and it's next are not equal to last
    // when fast_ptr == last, linkedList is odd
    // when fast_ptr.next == last, linkedList is even
    // Note: this also checks if the list has 1 or 2 nodes only
    while( (fast_ptr != last) && (fast_ptr.next != last) ){
      slow_ptr = slow_ptr.next;
      fast_ptr = fast_ptr.next.next;
    }
    if(fast_ptr == last){
      System.out.println("Middle node: "+ slow_ptr.value);
    }else{
      System.out.println("Middle nodes: "+ slow_ptr.value + " and " + slow_ptr.next.value);
    }
    
  }

  public boolean hasLoop(){
    // Using Floyd's Cycle-Finding Algorithm

    Node slow_ptr = first;
    Node fast_ptr = first;

    while( (fast_ptr.next != null) && (fast_ptr.next.next !=null) ){
      slow_ptr = slow_ptr.next;
      fast_ptr = fast_ptr.next.next;

      if(slow_ptr == fast_ptr){
        return true;
      }
    }
    return false;
  }

  // creating a loop to test hasLoop() method
  public void createLoop(){
    Node new_node = new Node(-7);
    last.next = new_node;
    new_node.next = first;
  }
  
  // Find the kth node from the end of a linkedList in one pass
  // Strategy:
  // Use two pointers where the last pointer is 
  // k-1 from the first pointer
  // Move till you get to the end of the list and the node the
  // first pointer points to is the kth node
}
